collaborative-plonk/src/round5.rs

```diff
@@ -0,0 +1,429 @@
+        + MSMProvider<P::G2>,
+    P::ScalarField: mpc_core::traits::FFTPostProcessing,
+{
+    pub(super) driver: T,
+    pub(super) domains: Domains<P>,
+    pub(super) challenges: Round4Challenges<P>,
+    pub(super) proof: Round4Proof<P>,
+    pub(super) polys: FinalPolys<T, P>,
+    pub(super) data: PlonkData<T, P>,
+}
+pub(super) struct Round5Challenges<P: Pairing> {
+    beta: P::ScalarField,
+    gamma: P::ScalarField,
+    alpha: P::ScalarField,
+    xi: P::ScalarField,
+    v: [P::ScalarField; 5],
+}
+
+impl<P: Pairing> Round5Challenges<P> {
+    fn new(round4_challenges: Round4Challenges<P>, v: [P::ScalarField; 5]) -> Self {
+        Self {
+            beta: round4_challenges.beta,
+            gamma: round4_challenges.gamma,
+            alpha: round4_challenges.alpha,
+            xi: round4_challenges.xi,
+            v,
+        }
+    }
+}
+
+impl<P: Pairing + CircomArkworksPairingBridge> Round4Proof<P>
+where
+    P::BaseField: CircomArkworksPrimeFieldBridge,
+    P::ScalarField: CircomArkworksPrimeFieldBridge,
+{
+    fn into_final_proof(self, commit_wxi: P::G1, commit_wxiw: P::G1) -> PlonkProof<P> {
+        PlonkProof {
+            a: self.commit_a.into(),
+            b: self.commit_b.into(),
+            c: self.commit_c.into(),
+            z: self.commit_z.into(),
+            t1: self.commit_t1.into(),
+            t2: self.commit_t2.into(),
+            t3: self.commit_t3.into(),
+            eval_a: self.eval_a,
+            eval_b: self.eval_b,
+            eval_c: self.eval_c,
+            eval_s1: self.eval_s1,
+            eval_s2: self.eval_s2,
+            eval_zw: self.eval_zw,
+            wxi: commit_wxi.into(),
+            wxiw: commit_wxiw.into(),
+            protocol: "plonk".to_string(),
+            curve: P::get_circom_name(),
+        }
+    }
+}
+
+impl<T, P: Pairing> Round5<T, P>
+where
+    for<'a> T: PrimeFieldMpcProtocol<P::ScalarField>
+        + PairingEcMpcProtocol<P>
+        + FFTProvider<P::ScalarField>
+        + MSMProvider<P::G1>
+        + MSMProvider<P::G2>,
+    P::ScalarField: FFTPostProcessing,
+    P: CircomArkworksPairingBridge,
+    P::BaseField: CircomArkworksPrimeFieldBridge,
+    P::ScalarField: CircomArkworksPrimeFieldBridge,
+{
+    fn div_by_zerofier(
+        driver: &mut T,
+        inout: &mut Vec<FieldShare<T, P>>,
+        n: usize,
+        beta: P::ScalarField,
+    ) {
+        let inv_beta = beta.inverse().expect("Highly unlikely to be zero");
+        let inv_beta_neg = -inv_beta;
+
+        let mut is_one = inv_beta_neg.is_one();
+        let mut is_negone = inv_beta.is_one();
+
+        if !is_one {
+            for el in inout.iter_mut().take(n) {
+                if is_negone {
+                    *el = driver.neg(el);
+                } else {
+                    *el = driver.mul_with_public(&inv_beta_neg, el);
+                }
+            }
+        }
+
+        std::mem::swap(&mut is_negone, &mut is_one);
+
+        for i in n..inout.len() {
```

Comment:
This also seems slow and could use a dedicated impl at some point, but let's do some profiling first.

Reply:
it is a linear scan. I removed all the if else here since they should not come into effect for a normal run.